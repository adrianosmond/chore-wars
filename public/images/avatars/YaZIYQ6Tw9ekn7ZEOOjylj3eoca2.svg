<svg viewBox="0 0 465 465" xmlns="http://www.w3.org/2000/svg">
  <path fill="none" d="M-1-1h467v467H-1z"/>
  <path d="M127.008 175.641l-10.15 191.872s63 8.764 115.029 8.764c55.333 0 108.672-15.466 108.672-15.466l-8.347-179.751" fill="#ffcf00"/>
  <path d="M332.212 185.223c0 52.888-45.936 95.757-102.6 95.757-56.665 0-102.604-42.87-102.604-95.757v-4.163c0-52.89 45.939-95.764 102.604-95.764 56.664 0 102.6 42.875 102.6 95.764v4.163" fill="#ffcf00"/>
  <path d="M348.465 235.743c-.853 16.338-9.093 35.544-17.986 35.076-8.88-.47-15.054-20.432-14.2-36.774.866-16.341 8.44-22.873 17.32-22.404 8.893.47 15.733 7.76 14.866 24.102m-231.935 0c.862 16.338 9.1 35.544 17.987 35.076 8.888-.472 15.06-20.432 14.199-36.774-.861-16.341-8.432-22.873-17.317-22.404-8.89.47-15.728 7.76-14.868 24.102" fill="#fbd09a"/>
  <path d="M232.499 412.899c28.604 0 43.677-14.933 50.685-24.26-9.57-2.34-11.46-11.449-11.46-16.688v-22.078h-78.45v22.078c0 5.24-1.889 14.349-11.46 16.687 7.007 9.328 22.085 24.26 50.685 24.26" fill="#fcd5a7"/>
  <path d="M342.905 418.31c-11.293-22.147-49.545-27.374-59.402-29.586-.112-.02-.208-.058-.319-.086-7.008 9.328-22.081 24.26-50.685 24.26-28.6 0-43.678-14.932-50.686-24.26-.108.028-.205.066-.316.086-9.857 2.212-48.108 7.439-59.4 29.587-11.029 21.634-16.336 46.797-16.336 46.797H359.24s-5.307-25.163-16.334-46.797" fill="#999"/>
  <path d="M232.499 363.467c-28.182 0-69.547-21.402-82.386-44.227-12.84-22.828-19.618-70.617-19.618-111.399 0-63.522 45.666-102.002 102.004-102.002 56.338 0 102.006 38.48 102.006 102.002 0 40.782-6.773 88.57-19.626 111.399-12.834 22.825-54.198 44.227-82.379 44.227h-.001" fill="#fcd5a7"/>
  <path d="M164.09 164.19s-13.215 63.886-24.231 65.763c-11.014 1.876-20.564 0-20.564 0l10.45-67.226 18.071-25.7 24.353-23.495 69.467-19.492 75.923 38.583 22.373 77.1-5.24 20.23c-13.707-7.124-20.56-69.066-20.56-69.066l-3.253 53.09c-19.82-1.37-58.868-7.977-67.59-10.374-32.784-9.022-38.27-10.71-79.2-39.414" fill="#ffcf00"/>
  <path d="M183.101 238.993a9.542 9.542 0 0 1-9.542 9.542c-5.271 0-9.544-4.271-9.544-9.542a9.542 9.542 0 0 1 9.544-9.545c5.27 0 9.542 4.27 9.542 9.545m117.884 0a9.54 9.54 0 0 1-9.542 9.542c-5.27 0-9.543-4.271-9.543-9.542a9.542 9.542 0 0 1 9.543-9.545 9.54 9.54 0 0 1 9.542 9.545" fill="#211c1d"/>
  <path d="M213.132 284.72l19.284-42.967 19.287 42.967-19.287 6.364-19.284-6.364" fill="#edc093"/>
  <path d="M251.479 315.61c-4.431-1.357-8.252-3.037-12.776-3.037-4.528 0-6.203 2.636-6.203 2.636s-1.673-2.636-6.203-2.636c-4.525 0-8.344 1.68-12.777 3.038-4.436 1.356-6.887 1-6.887 1 1.004 6.85 8.646 15.905 25.867 15.905 17.225 0 24.865-9.055 25.867-15.905 0 0-2.452.356-6.888-1" fill="#f14dae"/>
  <path d="M251.479 315.61c-4.431-1.357-8.252-3.037-12.776-3.037-4.528 0-6.203 2.636-6.203 2.636s-1.673-2.636-6.203-2.636c-4.525 0-8.344 1.68-12.777 3.038-4.436 1.356-6.887 1-6.887 1s3.834 5.728 14.216 5.728c1.807 0 5.406-.963 7.218-.963 1.812 0 2.682 1.055 4.433 1.055 1.75 0 2.624-1.055 4.436-1.055 1.81 0 5.407.963 7.216.963 10.379 0 14.215-5.728 14.215-5.728s-2.452.356-6.888-1" fill="#f366ba"/>
</svg>
