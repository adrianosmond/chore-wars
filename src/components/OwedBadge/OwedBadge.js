import React from 'react';
import { connect } from 'react-redux';

import PopUpMenu from 'components/PopUpMenu';

import { paidDebt } from 'actions/pointActions';

const OwedBadge = ({
  player, idx, debtPaid, gameId, points,
}) => {
  const playerPoints = points[player.id];
  if (playerPoints.isOwed === 0) return null;
  return (
    <PopUpMenu extraClasses={`players__paid-button players__paid-button--${idx + 1}`}
      side={ idx === 0 ? 'left' : 'right' }
      options={[{
        type: 'button',
        text: `${player.name} was paid back`,
        onClick: () => debtPaid(player.id, gameId),
      }]}>
      <span className="players__player-owed" role="img" aria-labelledby={`${player.name} is owed`}>ðŸŒŸ</span>
      { playerPoints.isOwed > 1 ? <span className="players__owed-multiple">{playerPoints.isOwed}</span> : null }
    </PopUpMenu>
  );
};

const mapStateToProps = state => ({
  points: state.points.present,
  gameId: state.session.game.gameId,
});

const mapDispatchToProps = dispatch => ({
  debtPaid: (player, game) => dispatch(paidDebt(player, game)),
});

export default connect(mapStateToProps, mapDispatchToProps)(OwedBadge);
