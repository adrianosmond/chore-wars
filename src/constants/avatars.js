export const avatarOptions = {
  skinColor: [
    'Pale',
    'Light',
    'Tanned',
    'Yellow',
    'Brown',
    'DarkBrown',
    'Black',
  ],
  topType: [
    'NoHair',
    'Eyepatch',
    'Hat',
    'Hijab',
    'Turban',
    'WinterHat1',
    'WinterHat2',
    'WinterHat3',
    'WinterHat4',
    'LongHairBigHair',
    'LongHairBob',
    'LongHairBun',
    'LongHairCurly',
    'LongHairCurvy',
    'LongHairDreads',
    'LongHairFrida',
    'LongHairFro',
    'LongHairFroBand',
    'LongHairNotTooLong',
    'LongHairShavedSides',
    'LongHairMiaWallace',
    'LongHairStraight',
    'LongHairStraight2',
    'LongHairStraightStrand',
    'ShortHairDreads01',
    'ShortHairDreads02',
    'ShortHairFrizzle',
    'ShortHairShaggyMullet',
    'ShortHairShortCurly',
    'ShortHairShortFlat',
    'ShortHairShortRound',
    'ShortHairShortWaved',
    'ShortHairSides',
    'ShortHairTheCaesar',
    'ShortHairTheCaesarSidePart',
  ],
  hatColor: [
    'Black',
    'Blue01',
    'Blue02',
    'Blue03',
    'Gray01',
    'Gray02',
    'Heather',
    'PastelBlue',
    'PastelGreen',
    'PastelOrange',
    'PastelRed',
    'PastelYellow',
    'Pink',
    'Red',
    'White',
  ],
  hairColor: [
    'Auburn',
    'Black',
    'Blonde',
    'BlondeGolden',
    'Brown',
    'BrownDark',
    'PastelPink',
    'Platinum',
    'Red',
    'SilverGray',
  ],
  facialHairType: [
    'Blank',
    'BeardMedium',
    'BeardLight',
    'BeardMagestic',
    'MoustacheFancy',
    'MoustacheMagnum',
  ],
  facialHairColor: [
    'Auburn',
    'Black',
    'Blonde',
    'BlondeGolden',
    'Brown',
    'BrownDark',
    'Platinum',
    'Red',
  ],
  accessoriesType: [
    'Blank',
    'Kurt',
    'Prescription01',
    'Prescription02',
    'Round',
    'Sunglasses',
    'Wayfarers',
  ],
  clotheType: [
    'BlazerShirt',
    'BlazerSweater',
    'CollarSweater',
    'Hoodie',
    'Overall',
    'ShirtCrewNeck',
    'ShirtScoopNeck',
    'ShirtVNeck',
  ],
  clotheColor: [
    'Black',
    'Blue01',
    'Blue02',
    'Blue03',
    'Gray01',
    'Gray02',
    'Heather',
    'PastelBlue',
    'PastelGreen',
    'PastelOrange',
    'PastelRed',
    'PastelYellow',
    'Pink',
    'Red',
    'White',
  ],
};


export const editorOrder = [
  'skinColor',
  'topType',
  'hairColor',
  'hatColor',
  'facialHairType',
  'facialHairColor',
  'accessoriesType',
  'clotheType',
  'clotheColor',
];

export const labels = {
  skinColor: 'Skin Colour',
  topType: 'Hair or Headware',
  hairColor: 'Hair Colour',
  hatColor: 'Headware Colour',
  facialHairType: 'Facial Hair Type',
  facialHairColor: 'Facial Hair Colour',
  accessoriesType: 'Glasses',
  clotheType: 'Clothing Type',
  clotheColor: 'Clothing Colour',
};

export const isEditable = {
  skinColor: () => true,
  topType: () => true,
  hairColor: options => [
    'NoHair',
    'Eyepatch',
    'Hat',
    'Hijab',
    'Turban',
    'WinterHat1',
    'WinterHat2',
    'WinterHat3',
    'WinterHat4',
    'LongHairFrida',
    'LongHairShavedSides',
  ].indexOf(options.topType) < 0,
  hatColor: options => [
    'Hijab',
    'Turban',
    'WinterHat1',
    'WinterHat2',
    'WinterHat3',
    'WinterHat4',
  ].indexOf(options.topType) >= 0,
  facialHairType: options => options.topType !== 'Hijab',
  facialHairColor: options => isEditable.facialHairType(options) && options.facialHairType !== 'Blank',
  accessoriesType: () => true,
  clotheType: () => true,
  clotheColor: options => !options.clotheType.startsWith('Blazer'),
};

export const DefaultAvatar = {
  accessoriesType: 'Blank',
  avatarStyle: 'Transparent',
  clotheColor: 'Blue03',
  clotheType: 'BlazerShirt',
  eyeType: 'Default',
  eyebrowType: 'Default',
  facialHairColor: 'BrownDark',
  facialHairType: 'Blank',
  hairColor: 'BrownDark',
  mouthType: 'Twinkle',
  skinColor: 'Light',
  topType: 'LongHairStraight',
};
