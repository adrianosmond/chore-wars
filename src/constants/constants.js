const TIME_UNIT = 1000 * 60 * 60 * 24; // 1 day
const MIN_CHORE_FREQUENCY = 1;
const MAX_CHORE_FREQUENCY = 365;
const MIN_CHORE_POINTS = 1;
const MAX_CHORE_POINTS = 200;

const DATE_FORMAT = 'DD/MM/YYYY';

const MAX_POINT_DIFFERENCE = 500;

const MAX_NAME_LENGTH = 10;
const JOIN_CODE_LENGTH = 8;

const DefaultAvatar = {
  accessoriesType: 'Blank',
  avatarStyle: 'Transparent',
  clotheColor: 'Blue03',
  clotheType: 'BlazerShirt',
  eyeType: 'Default',
  eyebrowType: 'Default',
  facialHairColor: 'BrownDark',
  facialHairType: 'Blank',
  hairColor: 'BrownDark',
  mouthType: 'Twinkle',
  skinColor: 'Light',
  topType: 'LongHairStraight',
};

const ActionTypes = {
  addChore: 'ADD_CHORE',
  resetChoreDoneDate: 'RESET_CHORE_DONE_DATE',
  blockChore: 'BLOCK_CHORE',
  unblockChore: 'UNBLOCK_CHORE',
  removeChore: 'REMOVE_CHORE',
  updateChore: 'UPDATE_CHORE',
  setChores: 'SET_CHORES',
  makeChain: 'MAKE_CHAIN',
  breakChain: 'BREAK_CHAIN',
  addPoints: 'ADD_POINTS',
  claimPrize: 'CLAIM_PRIZE',
  paidDebt: 'PAID_DEBT',
  setPoints: 'SET_POINTS',
  updateUser: 'UPDATE_USER',
  setAuthUser: 'SET_AUTH_USER',
  setGame: 'SET_GAME',
  setPointsLoaded: 'SET_POINTS_LOADED',
  setChoresLoaded: 'SET_CHORES_LOADED',
  saveStatePostUndo: 'SAVE_STATE_POST_UNDO',
};

export {
  TIME_UNIT,
  MIN_CHORE_FREQUENCY,
  MAX_CHORE_FREQUENCY,
  MIN_CHORE_POINTS,
  MAX_CHORE_POINTS,
  MAX_NAME_LENGTH,
  JOIN_CODE_LENGTH,
  DATE_FORMAT,
  MAX_POINT_DIFFERENCE,
  ActionTypes,
  DefaultAvatar,
};
