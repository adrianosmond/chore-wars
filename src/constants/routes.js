export const LOGIN = '/'
export const CHORES = '/chores'
export const NEW_CHORE = '/chores/new'
