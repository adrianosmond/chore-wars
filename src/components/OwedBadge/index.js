import OwedBadge from './OwedBadge';

export default OwedBadge;
