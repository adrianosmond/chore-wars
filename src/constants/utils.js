import { TIME_UNIT } from './constants'
const sortByCurrentPoints = (a, b) => b.currentPoints - a.currentPoints

const processChore = (chore) => {
  const timeSinceChore = (new Date().getTime() - chore.lastDone) / TIME_UNIT
  const timeRemaining = timeSinceChore - chore.frequency
  const percentage = Math.min(100 * timeSinceChore / chore.frequency, 100)
  const multiplier = 1 + (1 / chore.frequency)
  const due = chore.lastDone + (chore.frequency * TIME_UNIT)
  return {
    currentPoints: Math.round(percentage / 100 * chore.pointsPerTime * Math.pow(multiplier, timeRemaining)),
    percentage,
    due
  }
}

export { sortByCurrentPoints, processChore }
