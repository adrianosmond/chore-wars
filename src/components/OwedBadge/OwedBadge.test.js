import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { shallow } from 'enzyme';
import { DefaultAvatar } from 'constants/avatars';
import OwedBadge from './OwedBadge';

const mockStore = configureMockStore([thunk]);
const session = {
  game: {
    gameId: 'test-game',
  },
};

describe('OwedBadge', () => {
  describe('With player not being owed', () => {
    it('Renders nothing for p1', () => {
      const player = {
        id: 'player1',
        avatar: DefaultAvatar,
        name: 'Player 1',
      };
      const store = mockStore({
        points: {
          present: {
            player1: {
              isOwed: 0,
            },
          },
        },
        session,
      });
      const badgeComponent = shallow(<OwedBadge store={store} player={player} idx={0} />).dive();
      expect(badgeComponent).toMatchSnapshot();
    });

    it('Renders nothing for p2', () => {
      const player = {
        id: 'player2',
        avatar: DefaultAvatar,
        name: 'Player 2',
      };
      const store = mockStore({
        points: {
          present: {
            player2: {
              isOwed: 0,
            },
          },
        },
        session,
      });
      const badgeComponent = shallow(<OwedBadge store={store} player={player} idx={1} />).dive();
      expect(badgeComponent).toMatchSnapshot();
    });
  });

  describe('With player being owed', () => {
    it('Renders badge with p1 being owed 1', () => {
      const player = {
        id: 'player1',
        avatar: DefaultAvatar,
        name: 'Player 1',
      };
      const store = mockStore({
        points: {
          present: {
            player1: {
              isOwed: 1,
            },
          },
        },
        session,
      });
      const badgeComponent = shallow(<OwedBadge store={store} player={player} idx={0} />).dive();
      expect(badgeComponent).toMatchSnapshot();
    });

    it('Renders badges with p2 being owed 1', () => {
      const player = {
        id: 'player2',
        avatar: DefaultAvatar,
        name: 'Player 2',
      };
      const store = mockStore({
        points: {
          present: {
            player2: {
              isOwed: 1,
            },
          },
        },
        session,
      });
      const badgeComponent = shallow(<OwedBadge store={store} player={player} idx={1} />).dive();
      expect(badgeComponent).toMatchSnapshot();
    });

    it('Renders badges with p1 being owed > 1', () => {
      const player = {
        id: 'player1',
        avatar: DefaultAvatar,
        name: 'Player 1',
      };
      const store = mockStore({
        points: {
          present: {
            player1: {
              isOwed: 2,
            },
          },
        },
        session,
      });
      const badgeComponent = shallow(<OwedBadge store={store} player={player} idx={0} />).dive();
      expect(badgeComponent).toMatchSnapshot();
    });

    it('Renders badges with p2 being owed > 1', () => {
      const player = {
        id: 'player2',
        avatar: DefaultAvatar,
        name: 'Player 2',
      };
      const store = mockStore({
        points: {
          present: {
            player2: {
              isOwed: 2,
            },
          },
        },
        session,
      });
      const badgeComponent = shallow(<OwedBadge store={store} player={player} idx={1} />).dive();
      expect(badgeComponent).toMatchSnapshot();
    });

    it('Handles clicks', () => {
      const player = {
        id: 'player1',
        avatar: DefaultAvatar,
        name: 'Player 1',
      };
      const store = mockStore({
        points: {
          present: {
            player1: {
              isOwed: 1,
            },
          },
        },
        session,
      });
      const badgeComponent = shallow(<OwedBadge store={store} player={player} idx={0} />).dive();
      const paidDebtFn = jest.fn();
      badgeComponent.setProps({
        debtPaid: paidDebtFn,
      });
      // Click the option in the popup menu
      badgeComponent.prop('options')[0].onClick();
      expect(paidDebtFn).toHaveBeenCalledTimes(1);
    });
  });
});
