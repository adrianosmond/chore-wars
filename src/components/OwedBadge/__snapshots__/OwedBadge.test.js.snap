// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OwedBadge With player being owed Renders badge with p1 being owed 1 1`] = `
<PopUpMenu
  extraClasses="players__paid-button players__paid-button--1"
  options={
    Array [
      Object {
        "onClick": [Function],
        "text": "Player 1 was paid back",
        "type": "button",
      },
    ]
  }
  side="left"
>
  <span
    aria-labelledby="Player 1 is owed"
    className="players__player-owed"
    role="img"
  >
    ðŸŒŸ
  </span>
</PopUpMenu>
`;

exports[`OwedBadge With player being owed Renders badges with p1 being owed > 1 1`] = `
<PopUpMenu
  extraClasses="players__paid-button players__paid-button--1"
  options={
    Array [
      Object {
        "onClick": [Function],
        "text": "Player 1 was paid back",
        "type": "button",
      },
    ]
  }
  side="left"
>
  <span
    aria-labelledby="Player 1 is owed"
    className="players__player-owed"
    role="img"
  >
    ðŸŒŸ
  </span>
  <span
    className="players__owed-multiple"
  >
    2
  </span>
</PopUpMenu>
`;

exports[`OwedBadge With player being owed Renders badges with p2 being owed > 1 1`] = `
<PopUpMenu
  extraClasses="players__paid-button players__paid-button--2"
  options={
    Array [
      Object {
        "onClick": [Function],
        "text": "Player 2 was paid back",
        "type": "button",
      },
    ]
  }
  side="right"
>
  <span
    aria-labelledby="Player 2 is owed"
    className="players__player-owed"
    role="img"
  >
    ðŸŒŸ
  </span>
  <span
    className="players__owed-multiple"
  >
    2
  </span>
</PopUpMenu>
`;

exports[`OwedBadge With player being owed Renders badges with p2 being owed 1 1`] = `
<PopUpMenu
  extraClasses="players__paid-button players__paid-button--2"
  options={
    Array [
      Object {
        "onClick": [Function],
        "text": "Player 2 was paid back",
        "type": "button",
      },
    ]
  }
  side="right"
>
  <span
    aria-labelledby="Player 2 is owed"
    className="players__player-owed"
    role="img"
  >
    ðŸŒŸ
  </span>
</PopUpMenu>
`;

exports[`OwedBadge With player not being owed Renders nothing for p1 1`] = `""`;

exports[`OwedBadge With player not being owed Renders nothing for p2 1`] = `""`;
